### PR TITLE
Mainstay SMT root commitment submission/retrieval and database integration, server spawn error handling

### PR DESCRIPTION
**Pull request by lawlawlaw**
* Interface to the mainstay API, and submit SMT root as a commitment to a mainstay slot whenever updated 
* Retrieval of the above as needed and submission to the database
* Handling server spawn errors  
* Update wallet with mainstay-confirmed proofs on request

### CODE REVIEW

File: shared/src/mainstay.rs [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+use std::error;
+use rocket::response::Responder;
+use rocket::http::{ Status, ContentType };
+use std::io::Cursor;
+
+type Result<T> = std::result::Result<T, Box<dyn error::Error> >;
+
+pub type Hash = monotree::Hash;
+
+type Payload<'a> = HashMap::<&'a str,&'a str>;
```
Comment (tomossomot) [inline shared/src/mainstay.rs:21]: `type Payload<'a> = HashMap<&'a str,&'a str>;`

This compiles but my IDE doesnt like it.

**Comment by tomossomot** [comment]
Im also getting test failure on `test_db_commitment_info_update`

File: shared/src/mainstay.rs (hunk 1 above)
Comment (lawlawlaw) [reply shared/src/mainstay.rs:21]: What does your IDE say about it?

**Comment by lawlawlaw** [comment]
> Looks good but there are lots of unused imports.

Some of those are actually used but the compiler does not think so e.g. serial_test. I plan to use some of them soon.

File: shared/src/mainstay.rs (hunk 1 above)
Comment (tomossomot) [reply shared/src/mainstay.rs:21]: Should be 
`type Payload<'a> = HashMap<&'a str,&'a str>;`
not 
`type Payload<'a> = HashMap::<&'a str,&'a str>;`

I dont think the latter isnt correct syntax

**Comment by tomossomot** [comment]
> > Looks good but there are lots of unused imports.
> 
> Some of those are actually used but the compiler does not think so e.g. serial_test. I plan to use some of them soon.

If you add 

```
#[cfg(test)]
#[macro_use]
extern crate serial_test;
```

 it will let the compiler know that this crate is used in tests.

File: shared/src/mainstay.rs (hunk 1 above)
Comment (lawlawlaw) [reply shared/src/mainstay.rs]: Fixed it

**Comment by lawlawlaw** [comment]
> > > Looks good but there are lots of unused imports.
> > 
> > 
> > Some of those are actually used but the compiler does not think so e.g. serial_test. I plan to use some of them soon.
> 
> If you add
> 
> ```
> #[cfg(test)]
> #[macro_use]
> extern crate serial_test;
> ```
> 
> it will let the compiler know that this crate is used in tests.
> 
> Also would you mind adding `#[allow(unused_imports)]` before unused imports while they are unused.

Ok done

**Comment by lawlawlaw** [comment]
> > > Looks good but there are lots of unused imports.
> > 
> > 
> > Some of those are actually used but the compiler does not think so e.g. serial_test. I plan to use some of them soon.
> 
> If you add
> 
> ```
> #[cfg(test)]
> #[macro_use]
> extern crate serial_test;
> ```
> 
> it will let the compiler know that this crate is used in tests.
> 
> Also would you mind adding `#[allow(unused_imports)]` before unused imports while they are unused.

Unused imports removed

File: client/src/state_entity/mod.rs [hunk 2]
```diff
@@ -1,6 +1,7 @@
 // state entity client side interface
 
 pub mod deposit;
+pub mod confirm_proofs;
```
Comment (tomossomot) [inline client/src/state_entity/mod.rs:4]: File missing?

Comment (lawlawlaw) [reply client/src/state_entity/mod.rs:4]: Added

File: client/src/state_entity/confirm_proofs.rs [hunk 3]
```diff
@@ -14,41 +14,67 @@ use crate::wallet::wallet::Wallet;
...
-    
     let shim = wallet.client_shim.clone();
 
-    //Get the funding txid and the corresponding funnding txid
-    //for key in wallet.shared_keys_mutable() {
-    for key in &mut wallet.shared_keys {
-        println!("getting smt proof");
+    
+    // Verify proof key inclusion in SE sparse merkle tree
+    println!("getting confirmed smt root");
```
Comment (tomossomot) [inline client/src/state_entity/confirm_proofs.rs:44]: debug!()
there are a few println!()s in this file

File: client/src/state_entity/confirm_proofs.rs [hunk 4]
```diff
@@ -14,41 +14,67 @@ use crate::wallet::wallet::Wallet;
...
 use crate::state_entity::util::{verify_statechain_smt};
 use super::api::{get_smt_proof, get_confirmed_smt_root};
+use crate::utilities::requests;
+use shared_lib::structs::{ConfirmProofsMsg};
+
+
+
+/// Message to server initiating state entity protocol.
+/// Shared wallet ID returned
+pub fn session_init(wallet: &mut Wallet, shared_key_id: String) -> Result<String> {
```
Comment (tomossomot) [inline client/src/state_entity/confirm_proofs.rs:24]: What is this for?

File: server/src/server.rs [hunk 5]
```diff
@@ -15,25 +15,43 @@ use log4rs::config::{Appender, Config as LogConfig, Root};
...
-            electrum_server: settings.get("electrum_server").unwrap().to_string(),
-            network: settings.get("network").unwrap().to_string(),
-            testing_mode: bool::from_str(settings.get("testing_mode").unwrap()).unwrap(),
-            fee_address,
-            fee_deposit: settings.get("fee_deposit").unwrap().parse::<u64>().unwrap(),
-            fee_withdraw: settings.get("fee_withdraw").unwrap().parse::<u64>().unwrap(),
-            block_time: settings.get("block_time").unwrap().parse::<u64>().unwrap(),
-            batch_lifetime: settings.get("batch_lifetime").unwrap().parse::<u64>().unwrap(),
-            punishment_duration: settings.get("punishment_duration").unwrap().parse::<u64>().unwrap(),
+        //mainstay_config is optional
```
Comment (tomossomot) [inline server/src/server.rs:27]: Maybe set to `mainstay::Config::from_test()` if `testing_mode` is true, `match settings.get("mainstay_config")` otherwise

Comment (lawlawlaw) [reply server/src/server.rs:30]: Ok, done

**Pull request merged**